pkg/binding/bindingpolicy.go

```diff
@@ -96,27 +103,26 @@ func (c *Controller) handleBindingPolicy(ctx context.Context, obj runtime.Object
 		return c.requeueWorkloadObjects(ctx, bindingPolicy.Name)
 	}
 
-	// handle deletion of bindingpolicy
-	return c.handleBindingPolicyDeletion(ctx, bindingPolicy)
+	// we delete finalizer if the policy is being deleted (not yet deleted).
+	if err = c.handleBindingPolicyFinalizer(ctx, bindingPolicy); err != nil {
+		return fmt.Errorf("failed to handle finalizer for bindingPolicy %s: %w", bindingPolicy.Name, err)
+	}
+
+	return c.updateResolutionForDeletedBindingPolicy(ctx, objIdentifier.ObjectName.Name)
 }
 
-func (c *Controller) handleBindingPolicyDeletion(ctx context.Context, bindingPolicy *v1alpha1.BindingPolicy) error {
-	if c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicy.GetName()) {
+func (c *Controller) updateResolutionForDeletedBindingPolicy(ctx context.Context, bindingPolicyName string) error {
```

Comment:
FYI: The word "Deleted" in this method's name is not exactly right --- the object has not actually been deleted yet, right?